Pool Prioritization

gshively11 opened the pull request:
- Added optional pool prioritization functionality.
- Fixed bug in heal process that was preventing heal failures from being detected properly in some cases.

If pool prioritization is enabled, DSNs are assigned a priority indicating their order of preference (i.e. a DSN for a local database might have a higher priority than a DSN for a remote failover database). A check will be run at a configureable interval to see if the active pool's priority is exceeded by any others in the collection. If this is the case, any pools with a higher priority will be healed (if applicable) and placed in priority order above the active pool.

This is a useful behavior if you want to automatically fail back to a preferred DSN when it becomes available again after previously failing over to a lower priority DSN.

gshively11 committed:
- Added optional pool prioritization functionality.
- Fixed bug in heal process that was preventing heal failures from being detected properly in some cases.

mayank23 commented on src/connection-pool-party.js:
```diff
@@ -484,4 +516,86 @@ export default class ConnectionPoolParty extends EventEmitter {
+        this.pools.sort(poolPrioritySort);
+        debug('prioritized pools');
+      })
+      .then(() => {
```
Promise.finally for  ` this._prioritizePromise = null;` ?

gergelyke replied:
afaik Promise.finally just landed in v8 6.3, which is not part of Node.js 8

gshively11 replied:
yea, using native promises so no `finally` in older versions of node

mayank23 replied:
But the project is transpiled with Babel, probably includes the es6 promise polyfill

gshively11 replied:
good point, i'll see if it's supported

gshively11 replied:
didn't seem to like it, i'll just keep the then/catch duplication for now

gergelyke commented on src/connection-pool-party.js:
```diff
@@ -474,6 +500,12 @@ export default class ConnectionPoolParty extends EventEmitter {
+            // some connection pool factories may opt to return an error instead
+            // of rejecting the promise. the existence of an error indicates that
+            // the pool did not heal successfully
+            return pool.error;
```
shouldn't you throw here?

gshively11 replied:
nope, returning is intended. `_healPool` should never reject, only resolve with errors, which is what `_healPools` is expecting

gergelyke commented on test/prioritized-pools.test.js:
```diff
@@ -0,0 +1,261 @@
+      warmupStrategy: serialWarmupStrategy,
+    });
+    return connection.warmup()
+      .then(() => {
```
probably you can make your tests a bit easier to read by adding async-await

gshively11 replied:
agreed, i'm planning on doing a complete refactor of this library for async/await.  i wanted to keep everything consistent until that point though.

chyipin commented on src/connection-pool-party.js:
```diff
@@ -194,12 +212,15 @@ export default class ConnectionPoolParty extends EventEmitter {
     .catch((err) => {
       this.emit('error', err);
+    })
+    .then(() => {
```
Why delaying the cleanup after `emit`-ing the `err`?  Seems like it is better to `cleanup` before `emit`.

gshively11 replied:
was just trying to reduce duplication, since i had to do the cleanup in a `then` before the `catch` and also in the `catch`.  do you see any issue with this shortcut?

chyipin replied:
I'm just concerned about the consumer taking action on the `err` event and operating on the pool before we finish cleaning things up.

gshively11 replied:
i suppose that's a possibility, i'll revert

chyipin commented on src/pool-priority-sort.js:
```diff
@@ -0,0 +1,9 @@
+export default function poolPrioritySort(a, b) {
+  if (a.dsn.priority > b.dsn.priority) {
```
You can simplify the entire function to:
`return a.dsn.priority - b.dsn.priority;`
It is the >0, <0, and ==0 that matters for sorting.

gshively11 replied:
neat, will update

mayank23 replied:
What if the dsn doesn’t have a priority ? Then you have undefined - undefined which is NaN. With explicit less and or greater than checks it would have been 0

chyipin commented on src/connection-pool-party.js:
```diff
@@ -484,4 +516,86 @@ export default class ConnectionPoolParty extends EventEmitter {
+      debug('_prioritizePools called during heal, skipping this run');
+      return;
+    }
+    if (this.pools.length === 0) {
```
We can move the `pools.length` check into `_startPrioritizingPools`. We can prevent starting the timer if we don't have more than 1 pool.

gshively11 replied:
i'd rather keep the timer going on the off chance that somehow a warmup happened without actually having any pools, since `_startPrioritizingPools` only starts after a warmup.  i can't think of a situation where we'd have 0 pools immediately after a warmup, but it seems safer to just let the timer run and have the handler respond appropriately to the pool size

chyipin replied:
We also don't need to have the timer for pool size of 1.
But I'm ok w/ this.

gshively11 replied:
yea, presumably if there's only one DSN, they wouldn't enable the prioritize flag

chyipin commented on src/connection-pool-party.js:
```diff
@@ -484,4 +516,86 @@ export default class ConnectionPoolParty extends EventEmitter {
+        // skip the heal. If there are any unhealthy pools,
+        // we need to heal them before sorting.
+        if (unhealthyPools.length === 0) {
+          this.pools.sort(poolPrioritySort);
```
We don't need to sort here, it will be sorted in the following `.then`.

gshively11 replied:
i believe that `_healPools` will return false if there are no unhealthy pools passed to it, which would result in an early return without a sort in the following `then`, so we do need it here

chyipin replied:
Well, we are returning `true` in the next two line; which will set `anyHealthyPools` to `true` in the following `then`.

gshively11 replied:
oops, you're right, i think i got part way through refactoring to a single sort and forgot to take this part out

gshively11 replied:
oops, you're right, i think i got part way through refactoring to a single sort and forgot to take this part out

chyipin commented on test/prioritized-pools.test.js:
```diff
@@ -0,0 +1,261 @@
+        },
+        {
+          user: 'sa',
+          password: 'PoolPartyyy9000',
```
We need another test w/ 3 dsns and wrong password here.  With the current implementation of the sorting ignoring failed healing, it will put a disconnected pool first.

chyipin commented on src/connection-pool-party.js:
```diff
@@ -484,4 +516,86 @@ export default class ConnectionPoolParty extends EventEmitter {
+        // if any of the eligible pools are healthy, we prioritize. it would be preferable
+        // if we identified which pools didn't heal and ignore them in the
+        // prioritization, but that would require a little refactoring.
+        this.pools.sort(poolPrioritySort);
```
Don't we track the pool connection state (`pool.connection.connected`)?

gshively11 replied:
you're right, we could just tweak the sort comparer to honor `pool.connection.connected`, good idea

gshively11 committed:
Updated based on PR feedback
- Made the priority sort respect connection health
- Simplified the priority sort logic
- Added extra test cases for the new sort logic

chyipin commented on src/pool-priority-sort.js:
```diff
@@ -0,0 +1,13 @@
+  if (!a.connection.connected && b.connection.connected) {
+    return 1;
+  }
+  if (!a.connection.connected && !b.connection.connected) {
```
It is more conventional to not have this compare and return.  We would also get the benefit of sorting the disconnected pools in priority.

gshively11 committed:
Addressed PR feedback
- Reverted small DRY change due to possible edge case

chyipin commented:
It would be good to add some documentation on `priority` in config.  e.g. starts from 0 and is the highest priority.

gshively11 committed:
Addressed PR feedback
- Removed unnecessary sort in prioritize logic

gshively11 committed:
Added explanation of priority value to API docs
